cmd/kubeadm/app/phases/controlplane/manifests_linux.go

```diff
@@ -0,0 +1,151 @@
+
+// runComponentAsNonRoot updates the pod manifest and the hostVolume permissions to run as non root.
+func runComponentAsNonRoot(componentName string, pod *v1.Pod, usersAndGroups *users.UsersAndGroups, cfg *kubeadmapi.ClusterConfiguration) error {
+	switch componentName {
+	case kubeadmconstants.KubeAPIServer:
+		return runKubeAPIServerAsNonRoot(pod, usersAndGroups, cfg, users.UpdatePathOwnerAndPermissions)
+	case kubeadmconstants.KubeControllerManager:
+		return runKubeControllerManagerAsNonRoot(pod, usersAndGroups, cfg, users.UpdatePathOwnerAndPermissions)
+	case kubeadmconstants.KubeScheduler:
+		return runKubeSchedulerAsNonRoot(pod, usersAndGroups, users.UpdatePathOwnerAndPermissions)
+	}
+
+	return errors.New(fmt.Sprintf("component name %q is not valid", componentName))
+}
+
+// runKubeAPIServerAsNonRoot updates the pod manifest and the hostVolume permissions to run kube-apiserver as non root.
+func runKubeAPIServerAsNonRoot(pod *v1.Pod, usersAndGroups *users.UsersAndGroups, cfg *kubeadmapi.ClusterConfiguration, updatePathOwnerAndPermissions pathOwerAndPermissionsUpdaterFunc) error {
+	uid := usersAndGroups.Users.ID(kubeadmconstants.KubeAPIServerUserName)
+	gid := usersAndGroups.Groups.ID(kubeadmconstants.KubeAPIServerUserName)
+	saPublicKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPublicKeyName)
+	if err := updatePathOwnerAndPermissions(saPublicKeyFile, *uid, *gid, 0600); err != nil {
+		return err
+	}
+
+	saPrivateKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.ServiceAccountPrivateKeyName)
+	saPrivateKeyGid := usersAndGroups.Groups.ID(kubeadmconstants.ServiceAccountKeyReadersGroupName)
+	if err := updatePathOwnerAndPermissions(saPrivateKeyFile, 0, *saPrivateKeyGid, 0640); err != nil {
+		return err
+	}
+
+	apiServerKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKeyName)
+	if err := updatePathOwnerAndPermissions(apiServerKeyFile, *uid, *gid, 0600); err != nil {
+		return err
+	}
+
+	apiServerKubeletClientKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerKubeletClientKeyName)
+	if err := updatePathOwnerAndPermissions(apiServerKubeletClientKeyFile, *uid, *gid, 0600); err != nil {
+		return err
+	}
+
+	frontProxyClientKeyName := filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyClientKeyName)
+	if err := updatePathOwnerAndPermissions(frontProxyClientKeyName, *uid, *gid, 0600); err != nil {
+		return err
+	}
+
+	if cfg.Etcd.External == nil {
+		apiServerEtcdClientKeyFile := filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerEtcdClientKeyName)
+		if err := updatePathOwnerAndPermissions(apiServerEtcdClientKeyFile, *uid, *gid, 0600); err != nil {
+			return err
+		}
+	}
+
+	pod.Spec.Containers[0].SecurityContext = &v1.SecurityContext{
+		Capabilities: &v1.Capabilities{
+			Drop: []v1.Capability{"ALL"},
+			Add:  []v1.Capability{"NET_BIND_SERVICE"},
```

Comment:
i think all CAP* changes should have comments on top of them.

Reply:
Done.